pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/migrating/xml/ForLoopCanBeForeach.xml

```diff
@@ -225,4 +226,42 @@
         ]]></code>
     </test-code>
 
+    <test-code>
+        <description>Consider iterators only if safe</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+        class Foo {
+          void loop() {
+            Iterable<DataFlowNode> path = null;
+
+            for (Iterator<DataFlowNode> i = path.iterator(); i.hasNext();) {
+              DataFlowNode inode = i.next();
+              path.remove(inode); // throws ConcurrentModificationException if it were a foreach
+              if (inode.getVariableAccess() == null) {
+                continue;
+              }
+            }
+          }
+        }
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Do not report iterator loop if we can't find iterator decl</description>
```

Comment:
Hmm yeah I think I intended it as "don't report if we can't find *iterator* decl" though the test code addresses a missing *iterable* déclaration 
I think there might be a few edge cases which would mess with the rule in that case. For instance, if the iterator is declared as a field in a parent class (symbol table wouldn't be able to resolve it) and `next ` was already called as a side effect eg of a constructor. Then the loop would not loop through all elements of the iterate.  Might be purely theoretical though